sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/xml/XmlInferSchemaSuite.scala

```diff
@@ -0,0 +1,296 @@
+            |            <element>2</element>
+            |            <element>3</element>
+            |            <element>4</element>
+            |         </array>""".stripMargin,
+              """<object>
+            |            <field>str</field>
+            |         </object>""".stripMargin
+            )
+          )
+        ),
+        Seq(Row(214748364700L), Row(1)),
+        null
+      ) ::
+      Row(null, null, Seq("""<field>str</field>""", """<field>1</field>""")) ::
+      Row(null, null, Seq("1", "2", "3")) :: Nil
+    )
+  }
+
+  test("Handling missing fields") {
+    val xmlDF = readData(missingFields)
+
+    val expectedSchema = StructType(
+      StructField("a", BooleanType, true) ::
+      StructField("b", LongType, true) ::
+      StructField("c", ArrayType(LongType, true), true) ::
+      StructField("d", StructType(StructField("field", BooleanType, true) :: Nil), true) ::
+      StructField("e", StringType, true) :: Nil
+    )
+
+    assert(expectedSchema === xmlDF.schema)
+
+  }
+
+  test("Complex field and type inferring") {
+    val xmlDF = readData(complexFieldAndType1, Map("prefersDecimal" -> "true"))
+    val expectedSchema = StructType(
+      StructField(
+        "arrayOfArray1",
+        ArrayType(StructType(StructField("item", ArrayType(StringType, true)) :: Nil)),
+        true
+      ) ::
+      StructField(
+        "arrayOfArray2",
+        ArrayType(StructType(StructField("item", ArrayType(DecimalType(21, 1), true)) :: Nil), true)
+      ) ::
+      StructField("arrayOfBigInteger", ArrayType(DecimalType(21, 0), true), true) ::
+      StructField("arrayOfBoolean", ArrayType(BooleanType, true), true) ::
+      StructField("arrayOfDouble", ArrayType(DoubleType, true), true) ::
+      StructField("arrayOfInteger", ArrayType(LongType, true), true) ::
+      StructField("arrayOfLong", ArrayType(DecimalType(20, 0), true), true) ::
+      StructField("arrayOfNull", ArrayType(StringType, true), true) ::
+      StructField("arrayOfString", ArrayType(StringType, true), true) ::
+      StructField(
+        "arrayOfStruct",
+        ArrayType(
+          StructType(
+            StructField("field1", BooleanType, true) ::
+            StructField("field2", StringType, true) ::
+            StructField("field3", StringType, true) :: Nil
+          ),
+          true
+        ),
+        true
+      ) ::
+      StructField(
+        "struct",
+        StructType(
+          StructField("field1", BooleanType, true) ::
+          StructField("field2", DecimalType(20, 0), true) :: Nil
+        ),
+        true
+      ) ::
+      StructField(
+        "structWithArrayFields",
+        StructType(
+          StructField("field1", ArrayType(LongType, true), true) ::
+          StructField("field2", ArrayType(StringType, true), true) :: Nil
+        ),
+        true
+      ) :: Nil
+    )
+    assert(expectedSchema === xmlDF.schema)
+  }
+
+  test("complex arrays") {
+    val xmlDF = readData(complexFieldAndType2)
```

Comment:
Is the omission of `arrayOfStruct` and `complexArrayOfStruct` intentional?

Reply:
Yes. We converted the array of arrays into arrays of struct as XML doesn't support arrays of arrays. I will remove these for clarity.